### PR TITLE
add some fun to flux 🎉️🪅️

### PR DESCRIPTION
**Pull request by vsoch**
Problem: we do not have enough fun in Flux.
Solution: we can take advantage of the "bug" that we can add commands to Flux and have them not show up. "The bug is a feature!" This should add a bit of hidden fun that we can tweak over the years to add to the surprise. :)

### CODE REVIEW

**Comment by garlick** [comment]
Ooh, this reminds me of the old [fortune(6)](https://linux.die.net/man/6/fortune) program.  Hey could we call it `flux-fortune` and have it print random pearls of flux wisdom from a database somewhere, and the occasional surprise ascii art present?

**Comment by vsoch** [comment]
oh totally - can we keep that ascii art? You are saying you want the ascii art to be presented at random (with what probability?) for some fortunes, or presented as one of the fortunes to be chosen from randomly? I would say we could mirror lolcow and have the graphic with the ascii at the bottom, but I do like the idea that the ascii art is one level deeper in the command.

For the "database" to start - we probably can just do a simple list and random.choice (don't need anything like a real database!)

**Comment by garlick** [comment]
No real opinions here, one self contained python program sounds fine, random selection from a flat list sounds fine!

**Comment by vsoch** [comment]
okay nice - I'm almost done! So - I've kind of done this in the past - I have a list of some "wisdom" I've written plus once I've found I like. I tried to remove the inappropriate ones, hopefully you won't regret this! :laughing: This is a cool approach because we can have the entire team contribute their favorite fortunes (that they have made up or found).

**Comment by garlick** [comment]
Hopefully fluxish stuff?  I mean we should probably try to keep it sort of on topic :-)

**Comment by vsoch** [comment]
Too late :)

**Comment by vsoch** [comment]
Feel free to nit pick the ones that you think are OK to add. I was thinking it would be OK to have some random funny ones (akin to fortune) but of course I don't want to cross into territory of inappropriate or crass, etc. Let me know which in the list are ok and I'll filter it down to those.

**Comment by garlick** [comment]
Oh my :-) 

I don't think this can be a general fortune database.   Not to be a stick in the mud (back reference!) but could we drop all the non-flux entries?  If we don't limit the scope then we'll put ourselves in a position where we are crowd sourcing and curating a general list of aphorisms, which I don't think we're qualified or have time to do.

**Comment by vsoch** [comment]
haha sure, that's most of them but we will need to build it up then! I'll keep the ones that are marginally related to computers (e..g, the modem one).

**Comment by vsoch** [comment]
okay left a small set that are marginally computer / hpc / robot related - will try to think of better ones over dinner!

**Comment by garlick** [comment]
I see you managed to get an :avocado: in there :-)

**Comment by vsoch** [comment]
> I see you managed to get an avocado in there :-)

That's a true story!  (from maybe 2016?) I eat dinner in front of my computer and I have like, 4 avocados at the end, and I reached for my mouse but I grabbed an avocado instead. It was hilarious so I wrote it down. I already have a few more good ones (related to flux) will add after dinner!

**Comment by garlick** [comment]
What do you think of "Fortunes should not be limited to ƒ*ing ASCII"?

It's probably even funnier if your character set is misconfigured.

**Comment by vsoch** [comment]
> What do you think of "Fortunes should not be limited to ƒ*ing ASCII"?

I have a fairly high tolerance for swear words (or implicated ones) so it would be ok with me, but probably best to check with others! There was a pun I saw "I don't give a flux!" but I didn't think that was appropriate, haha.

**Comment by garlick** [comment]
Well it's only a faint whiff of a swear word.  OK, I'm signing off for the evening.

**Comment by garlick** [comment]
Yeah never mind on my suggestion from last night.  Channeling our frustrations is probably not a great use of this (though I'm sure it motivated some of the stuff in the old unix fortune(6) database)

Two ideas for desirable categories here are "flux one liners" (similar to [awk one liners](http://tuxgraphics.org/~guido/scripts/awk-one-liner.html), and a "did you know?" format, like 
- Did you know that a local Flux instance can be started anywhere with `flux start`?
- Like `top(1)`?  Try `flux-top(1)`!

**Comment by grondo** [comment]
The command could also sometimes display a random man page (if stdout is a tty)

**Comment by vsoch** [comment]
@grondo but are the manpages fun? Wouldn't that be a terrible fortune? :laughing: 

I'd be okay adding that, but (akin to the ascii) maybe just have it happen a tiny percentage of the time (1%)

**Comment by garlick** [comment]
The man pages might be a bit long for this.  fortune(6) used to be something people put in their .profile so the get a new tidbit at each login.

Another thought: valentine poems that only display in the lead-up to valentine's day?  We already had some good ones!

**Comment by vsoch** [comment]
oh totally! I could actually just check the date time to get the month, let me add a few of those now. They will be a surprise for next year!

**Comment by grondo** [comment]
Sorry I was aware of `fortune` but hadn't quite put it together that this was a strict `fortune` clone. (I was perhaps thinking of something that gives a random tip or bit of key knowledge for newer users)

It occurs to me this might be better suited to a framework project and not necessarily a flux-core command? That way the database of one-liners could be updated without having to reinstall flux-core?

Anyway, here's another one

 * Did you know "fuzzybunny" is a valid Flux jobid?

**Comment by garlick** [comment]
fuzzybunny :rabbit: love it!

Sorry, I guess this is whatever we want it to be.  If we want to display random man pages then it's fine by me.

Seems like this could be in flux-core until we need to spin it off with its own maintainer because it's attracting a pestilence of pull requests :-)

**Comment by vsoch** [comment]
> Seems like this could be in flux-core until we need to spin it off with its own maintainer because it's attracting a pestilence of pull requests :-) Edit: doing so would somewhat undermine its "easter egg" nature.

This is problem I want flux to have :)

**Comment by grondo** [comment]
No, given the extra information the man page doesn't make sense. This is not a teaching tool about Flux but just another case of developers being cute :smiley_cat:. I was merely explaining myself after being shot down.

 Doesn't hurt to keep it here, but I figured we may not want to apply the same testing and commit message rigor here as we would do with functionality.

**Comment by garlick** [comment]
I think I'd like to see the "purpose" of this to ultimately be to dispense random tidbits of flux knowledge that will educate in a short attention span kind of way.  I think mixing in random fun stuff is OK but maybe we should shoot for an 80/20 rule on useful/fun?

If we're in agreement then I'd like to see that goal added as a comment in the source code so those that come along later will have an idea of the ground rules for contribution.   Maybe we could enforce that the size of the "useful" array is 5* size of the "fun" array and print a "fun overload"  error if not, then add a CI test that runs `flux fortune` ensures it doesn't fail.

What do you think @vsoch and @grondo?

**Comment by vsoch** [comment]
This might be inspiration for a future "generate more graphics with simple colors and terminal characters" https://github.com/robole/fetching. I really like the space aliens one :space_invader:

**Comment by vsoch** [comment]
> If we're in agreement then I'd like to see that goal added as a comment in the source code so those that come along later will have an idea of the ground rules for contribution. Maybe we could enforce that the size of the "useful" array is 5* size of the "fun" array and print a "fun overload" error if not, 

I think that's fair! I spent all last night coming up with the "fun" list, so maybe we could bring together others to come up with the "useful stuff?" list? 

> then add a CI test that runs flux fortune ensures it doesn't fail.

Hmm I guess my only disagreement here is that I don't want it to obviously show up as a test. I think the module should be generally hard to find, and if it's provided as test, anyone that cds into that folder or looks at CI logs is likely to see it. My proposal is that for this first update we:

1. Add in enough "useful tips" so everyone here is satisfied
2. Write a contributor note to the top of the file
3. No automated testing until adding new lines has proven to be problematic (e.g., solve the problem when we have it and not before we do).

**Comment by garlick** [comment]
We'll want tests for this like anything else in flux-core.  Probably just a run added to `t/t0001-basic.t` would suffice though, and not testing the ratio is fine by me!

**Comment by vsoch** [comment]
okay will do after dinner!

**Comment by vsoch** [comment]
Aside from rebasing again (which likely could be a multi-time thing) what other changes would you like to this PR?

**Comment by garlick** [comment]
I don't think there's any rush - mergify will rebase when we finally set merge-when-passing and AFAICT this is a low risk for conflicts since it's all new.

Perhaps if/when you have time, you could add logic so it chooses an informational code path some configurable percentage of the time (80 was proposed), and then when I have time maybe I can come up with an initial set of informational array entries?

**Comment by vsoch** [comment]
> chooses an informational code path some configurable percentage of the time (80 was proposed), and then when I have time maybe I can come up with an initial set of informational array entries?

Yep, will do! I'll put dummy stuff for now.

**Comment by vsoch** [comment]
okay all set! 

- You can add facts to the "facts" list (I added a few)
- A new `--category` or `-c` flag allows you to choose a category
   ```bash
    flux fortune -c all            # this is the default
    flux fortune -c valentines     # show valentine fortune
    flux fortune -c art            # show art
    flux fortune -c facts          # show learning facts / tidbits
    flux fortune -c fun            # show fun fortune
   ```
 - A test added for each category and help
 - Feel free to populate more facts as you like!

I did a flux start in the new docker environment and it segfaulted, so I think I need to go back and check over if that's a result of the non-root base. I'll do that next time I develop something substantial - probably not a huge issue for now since I'm the only one using it. Sorry I missed it before (it could also be something transient because I don't recall this happening before).

**Comment by vsoch** [comment]
> Edit: it would be sort of nice to have command line options to force the category too (extra credit?)

You can send me avocados at:

```
The Vanessasaurus
1 Snowy Dinosaur Way
Dinosaur Mountains, Dinosaur Land, USA
```

File: src/bindings/python/flux/cli/fortune.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        parser.add_argument(
+            "-c",
+            "--category",
+            choices=["all", "valentines", "fun", "facts", "art"],
+            default="all",
+            help="Choose the category of fortunes to display.",
+        )
+        return parser
+
+    def generate_fortune(self, args):
```
[the remark this answers is not in the document]
Comment (vsoch) [reply src/bindings/python/flux/cli/fortune.py:72]: The returns are only used as a strategy to end the execution of the function where they are at, there is nothing returned in any case.

**Comment by garlick** [comment]
@vsoch - mind if I push some "flux facts" directly to your branch?

**Comment by vsoch** [comment]
Go for it! A flux master does not need to ask the padwan such questions!

**Comment by garlick** [comment]
huh, I can't seem to push there for some reason.  Well the commit in question is d70e199663fbdc015c69e1406b1ef7fd0cf38242 if you want to just pull that in from my fork of your branch (garlick add/flux-surprise).

And feel free to edit/improve.

**Comment by vsoch** [comment]
@garlick I think I just need to add you as a contributor - researchapps is an organization! Give me a minute...

**Comment by vsoch** [comment]
okay just gave you write! Sorry about the oversight!

**Comment by garlick** [comment]
@vsoch - I messed up python linting in the previous commit so I'm going to force push a redo.  Hope that's OK!  Then I'm done one this for today.

**Comment by vsoch** [comment]
Hmm some of the tests are failing I guess? And did you want to rebase or do you want me to?

**Comment by grondo** [comment]
I restarted that builder, which failed `libsubprocess/test_remote.t` with "Killed by signal 9", a known intermittent failure.

You don't have to do a rebase, mergify will do that for you when the merge-when-passing label is set.

File: configure.ac [hunk 2]
```diff
@@ -1,8 +1,9 @@
 ##
 # Prologue
 ##
-AC_INIT([flux-core],
-        m4_esyscmd([test -n "$FLUX_VERSION" && printf $FLUX_VERSION || git describe --always | awk '/.*/ {sub(/^v/, ""); printf "%s",$1; exit}']))
+AC_INIT([flux-core], 0.46.0)
```
Comment (grondo) [inline configure.ac:4]: This looks like a local change that was inadvertently committed.

**Comment by vsoch** [comment]
ohh noiiiice! okay, I'm good when you and @garlick are good then. I assume you sanity checking that the printing with colors / surrounding characters looks ok with the longer quotes? That's the only thing I'd want to check - my VScode is occupied running several terminals with minikube and testing otherwise I'd test now.

File: configure.ac (hunk 2 above)
Comment (vsoch) [reply configure.ac:4]: Yeah that's a mistake - that is the only way I can get it to build. Let me fix that and I'll just do the rebase too.

Comment (vsoch) [reply configure.ac]: okay fixed!

**Review by grondo — APPROVED** [review]
Ok, LGTM!

**Comment by grondo** [comment]
Oh, @vsoch I guess this one does need manual rebase since mergify does not have permission to rebase...

**Comment by vsoch** [comment]
I did rebase, so something else was since merged that beat this one.

**Comment by vsoch** [comment]
okay done again.

**Comment by grondo** [comment]
> I did rebase, so something else was since merged that beat this one.
Yeah, sorry. That's why we use mergify it will at least keep rebasing in the background for you if something else gets merged.

**Comment by vsoch** [comment]
> Yeah, sorry. That's why we use mergify it will at least keep rebasing in the background for you if something else gets merged.

Gotcha. Thanks!

**Pull request merged**